### PR TITLE
Fire tiles

### PR DESCRIPTION
**Pull request by nightblade9**

### CODE REVIEW

File: furusiya/model/components/fighter.py [hunk 1]
```diff
@@ -77,3 +77,5 @@ def die(self):
         death_function = self.death_function
         if death_function is not None:
             death_function(self.owner)
+        else:
```
Comment (nightblade9) [inline furusiya/model/components/fighter.py]: While the definition of `die` on `GameObject` is clever, it requires subclasses to call `super.die`. I think it makes more sense to think of `death_function` as an optional callback that, if present, gets called by `die`. I would recommend changing this.

File: furusiya/model/entities/fire.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        if self.turns_passed_alight >= config.data.enemies.fire.selfExtinguishTurns:
+            self.die()
+        if config.data.enemies.fire.spreadChance >= Game.random.randint(1, 100):
+            dx, dy = Game.random.choice([DELTA_UP, DELTA_DOWN, DELTA_LEFT, DELTA_RIGHT])
+            created_fire = Fire(self.x + dx, self.y + dy)
+            Game.area_map.entities.append(created_fire)
+
+    def die(self):
+        super().die()
+        Game.events.unbind('on_entity_move', self.on_entity_move)
```
Comment (nightblade9) [inline furusiya/model/entities/fire.py]: I don't like having to manually bind/unbind events when things die. We need to discuss the design of the event bus again. I think the message bus should just delegate to `GameObject` instances which have callbacks that are called on events; the bus shouldn't store the callbacks. (Here's a good reason why.)

Comment (nightblade9) [reply furusiya/model/entities/fire.py]: If I get a chance, I will implement this in a branch off of your branch.

File: furusiya/model/entities/fire.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class Fire(GameObject):
+    def __init__(self, x, y):
+        super().__init__(x, y, '*', 'Fire', colors.red, blocks=False)
+
+        Game.events.bind('on_entity_move', self.on_entity_move)
+        Game.events.bind('on_turn_pass', self.on_turn_passed)
+
+        self.turns_passed_alight = 0
```
Comment (nightblade9) [inline furusiya/model/entities/fire.py]: I would rather make this `turns_left_alight`. That way, if anything happens, we can increment the amount.

File: furusiya/model/entities/fire.py (hunk 2 above)
Comment (nightblade9) [inline furusiya/model/entities/fire.py]: I love the simplicity of this class. You can see at a glance what it is and how it works. It's fantastic.

File: furusiya/model/entities/fire.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            self.die()
+
+    def on_turn_passed(self):
+        self.turns_passed_alight += 1
+        if self.turns_passed_alight >= config.data.enemies.fire.selfExtinguishTurns:
+            self.die()
+        if config.data.enemies.fire.spreadChance >= Game.random.randint(1, 100):
+            dx, dy = Game.random.choice([DELTA_UP, DELTA_DOWN, DELTA_LEFT, DELTA_RIGHT])
+            created_fire = Fire(self.x + dx, self.y + dy)
+            Game.area_map.entities.append(created_fire)
```
Comment (nightblade9) [inline furusiya/model/entities/fire.py]: You forgot to check if that tile is a wall (fail) or an entity (which should probably catch fire and instantly extinguish). We need a test for this.

We also need a test for a case where we pick an empty ground tile (nothing should happen).

I would like a test for items (they should burn to ash), even if it fails now.

File: furusiya/model/entities/game_object.py [hunk 5]
```diff
@@ -94,3 +95,9 @@ def draw(self):
     def clear(self):
         # erase the character that represents this object
         Game.ui.con.draw_str(self.x, self.y, ' ', self.color)
+
+    def die(self):
+        Game.area_map.entities.remove(self)
+        self.clear()
+        self.name = ''
+        self.blocks = False
```
Comment (nightblade9) [inline furusiya/model/entities/game_object.py]: You know what would be cool? A map engulphed in flames and smothered in enemies. As the pillars of fire receed, you see dozens of red eyes peering at you from the darkness...

File: tests/furusiya/model/event/test_event_bus.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+import pytest
+
+from model.event.event_bus import EventBus
+
+
+class TestEventBus:
+    @pytest.fixture
+    def event_bus(self):
+        yield EventBus()
```
Comment (nightblade9) [inline tests/furusiya/model/event/test_event_bus.py]: I would stop writing fixtures for things that are trivial to construct, and just construct them inline in tests. It makes the test classes less noisy.

File: furusiya/model/entities/fire.py (hunk 2 above)
Comment (NegativeScript) [reply furusiya/model/entities/fire.py]: Now that I think about it, your approach of an event bus keeping a list of entities might be better.

All it takes to de-register a dead entity is an all-encompassing line of `event_bus.unregister(self)`, which could be incorporated into GameObject to make this seamless.

Thoughts?

**Comment by nightblade9** [comment]
General comment: I don't like how much we've mocked out -- I think we're almost over-mocking. I can't easily write a sensible check, such as "count the fire entities on the map and they should have increased by one."

Consolidated list of remaining changes.
- [x] config.json: spreadChance => spreadProbability
- [x] Fire should burn out as time passes
- [ ] `game.py`: self.events => self.event_bus
- [ ] `game_object` should have an `on_death` callback, not override `die`
- [ ] `game_object.die` should call `event_bus.unregister_me_plzkthx`
- [ ] fire test: if random spread targets a wall, nothing happens
- [ ] fire test: if random spread targets a an entity, it catches fire and instantly extinguishes
- [ ] fire test: if random spread targets an item, burns it to ash (new functionality)
